hal/src/stm32f2xx/i2c_hal.c

```diff
@@ -673,7 +689,7 @@ int32_t HAL_I2C_Request_Data_Ex(HAL_I2C_Interface i2c, const HAL_I2C_Transmissio
             /* SW Reset the I2C Peripheral */
             HAL_I2C_SoftwareReset(i2c);
             HAL_I2C_Release(i2c, NULL);
-            return 0;
+            return SYSTEM_ERROR_TIMEOUT;
```

Comment:
These changes will also break backwards compatibility. Old application and new DeviceOS will start returning absurdly large sizes.

Reply:
Given that the return type is `int32_t`, returning a negative value indicating an error wouldn't break the size range. Does it make sense? Or could you help me understand how it break the backwards compatibility.

Reply:
Sure, but what about `TwoWire::requestFrom()` ?

Reply:
Its _current_ implementation is just this:
```
size_t TwoWire::requestFrom(const WireTransmission& transfer) {
    auto conf = transfer.halConfig();
    return HAL_I2C_Request_Data_Ex(_i2c, &conf, nullptr);
}
```

And before the latest changes it was even worse:
```
  uint8_t result = HAL_I2C_Request_Data(_i2c, address, quantity, sendStop, NULL);
  return result;
```

A negative value would be cast into a `uint8_t` or `size_t` resulting in unexpected values for the application.

Again, this will happen for cases where the application was built for older DeviceOS and you have a system part running with these changes where we now return negative error codes.

Reply:
I just committed to revert the changes in spark_wiring_i2c.h and made proper change in spark_wiring_i2c.cpp. Kindly refer to the Files changed.

Reply:
This will not help existing applications built with old DeviceOS versions. We cannot change how things work in those.